src/platform/backends/hyperv_api/hcs_virtual_machine.cpp

```diff
@@ -0,0 +1,748 @@
+/*
+ * Copyright (C) Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <hyperv_api/hcs_virtual_machine.h>
+
+#include <hyperv_api/hcn/hyperv_hcn_create_endpoint_params.h>
+#include <hyperv_api/hcn/hyperv_hcn_wrapper.h>
+#include <hyperv_api/hcs/hyperv_hcs_compute_system_state.h>
+#include <hyperv_api/hcs/hyperv_hcs_event_type.h>
+#include <hyperv_api/hcs/hyperv_hcs_wrapper.h>
+#include <hyperv_api/hcs_plan9_mount_handler.h>
+#include <hyperv_api/hcs_virtual_machine_exceptions.h>
+#include <hyperv_api/virtdisk/virtdisk_snapshot.h>
+#include <hyperv_api/virtdisk/virtdisk_wrapper.h>
+
+#include <multipass/exceptions/formatted_exception_base.h>
+#include <multipass/top_catch_all.h>
+#include <multipass/virtual_machine_description.h>
+#include <multipass/vm_status_monitor.h>
+#include <platform/platform_win.h>
+#include <shared/windows/smb_mount_handler.h>
+
+#include <fmt/xchar.h>
+#include <scope_guard.hpp>
+
+#include <stdexcept>
+
+#include <WS2tcpip.h>
+
+namespace
+{
+
+/**
+ * Category for the log messages.
+ */
+constexpr auto log_category = "HyperV-Virtual-Machine";
+constexpr auto vhdx_filename = "current.vhdx";
+constexpr auto default_ssh_port = 22;
+
+namespace mpl = multipass::logging;
+namespace mpp = multipass::platform;
+using lvl = mpl::Level;
+using multipass::hyperv::hcn::HCN;
+using multipass::hyperv::hcs::HCS;
+using multipass::hyperv::virtdisk::VirtDisk;
+
+inline auto mac2uuid(std::string mac_addr)
+{
+    mac_addr.erase(std::remove(mac_addr.begin(), mac_addr.end(), ':'), mac_addr.end());
+    mac_addr.erase(std::remove(mac_addr.begin(), mac_addr.end(), '-'), mac_addr.end());
```

Comment:
This can be simpler now :slightly_smiling_face: 

```suggestion
    std::erase(mac_addr, ':');
    std::erase(mac_addr, '-');
```

Just a comment, no need to do it since you started with C++17.

Reply:
Done